torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -0,0 +1,213 @@
+  std::vector<torch::Tensor> tensors = message.tensors();
+  // append payload as a tensor
+  tensors.push_back(torch::from_blob(payload, payload_size, {torch::kChar}));
+  // append id and type as a tensor
+  tensors.push_back(torch::tensor(
+      {message.id(), (int64_t) message.type()}, {torch::kInt64}
+  ));
+
+  torch::save(tensors, os);
+}
+
+Message deserialize(std::istream& is) {
+  std::vector<torch::Tensor> tensors;
+
+  torch::load(tensors, is);
+
+  TORCH_CHECK(tensors.size() >= 2, "Failed to deserialize a message.");
+  auto miscTensor = std::move(tensors.back());
+  tensors.pop_back();
+  auto payloadTensor = std::move(tensors.back());
+  tensors.pop_back();
+
+  int64_t* miscItems = miscTensor.storage().data<int64_t>();
+  int64_t id = miscItems[0];
+  MessageType type = MessageType(miscItems[1]);
+
+  std::vector<char> payload(payloadTensor.numel());
+  std::memcpy(
+      payload.data(), payloadTensor.storage().data(), payloadTensor.numel());
+
+  return Message(std::move(payload), std::move(tensors), type, id);
+}
+
+} // namespace
+
+ProcessGroupAgent::ProcessGroupAgent(
+    std::string workerName,
+    std::unordered_map<std::string, int> nameMap,
+    std::shared_ptr<c10d::ProcessGroup> pg)
+    : RpcAgent(std::move(workerName), processRequestBlocking),
+      nameMap_(std::move(nameMap)),
+      stop_(false),
+      pg_(std::move(pg)),
+      nextId_(0) {
+  TORCH_CHECK(nameMap_.size() > 1, "ProcessGroupAgent requires world_size to "
+      "be at least 2, but got ", nameMap_.size());
+  auto workerRankIter = nameMap_.find(workerName_);
+  TORCH_CHECK(workerRankIter != nameMap_.end(), "Failed to resolve worker "
+      "name ", workerName_, " to a ProcessGroup rank.");
+  TORCH_CHECK(pg_->getRank() == workerRankIter -> second,
+      "Resolved worker rank ", workerRankIter -> second,
+      " does not match ProcessGroup rank ", pg_->getRank());
+
+  names_.resize(nameMap_.size());
+  for (auto& entry : nameMap_) {
+    names_[entry.second] = entry.first;
+  }
+  PythonRpcHandler::init();
+  sendThread_ = std::thread(&ProcessGroupAgent::sendLoop, this);
+  listenerThread_ = std::thread(&ProcessGroupAgent::listenLoop, this);
+}
+
+void ProcessGroupAgent::join() {
+  // Every process i sends a SHUTDOWN message to process i + 1. This is
+  // necessary for now because:
+  // 1. There is no abort API for ProcessGroup::recvAnysource yet. We have to
+  //    feed it a message or kill the thread.
+  // 2. A GLOO process cannot send message to itself. (there is an ongoing
+  //    effort to fix this problem).
+  pg_->barrier()->wait();
+  int dst = (pg_->getRank() + 1) % pg_->getSize();
+  enqueue(SendWork(dst, Message({}, {}, MessageType::SHUTDOWN)));
+  std::unique_lock<std::mutex> lock(sendQueueMutex_);
+  workConsumeCV_.wait(lock, [&] { return sendQueue_.empty(); });
+  stop_ = true;
+  lock.unlock();
+
+  workProduceCV_.notify_all();
+  sendThread_.join();
+  listenerThread_.join();
+
+  PythonRpcHandler::cleanUp();
```

Comment:
Can we move this into the destructor of ProcessGroupAgent.

Reply:
More generally, the lifetime of PythonRpcHandler's "static" variables actually seems directly tied to `ProcessGroupAgent` lifetime. If this is true, can we just make PythonRpcHandler a member of ProcessGroupAgent?

Reply:
that was initial thought, but turn out we need to call interface of PythonRpcHandler when we get result from future. At that point, future does not have processGroupAgent.